lib/http/request/proxy.rb

```diff
@@ -0,0 +1,65 @@
+module HTTP
+  class Request
+    class Proxy
+      def initialize(proxy_params, request_uri)
+        @proxy = proxy_params || find_proxy(request_uri)
```

Comment:
There are some caveats here that may need a decision:

- When it's user defined, it will ignore ENV defined values (maybe what we want is system to always override this?)
- If we want to define custom headers, it can't use any value from ENV

Should we try instead to merge everything? (So we can keep proxy defined by ENV but still set extra headers?)
What should be the precedence order? ENV always win?

Reply:
We definitely don't want ENV vars to take precedence over explicitly given proxy prams. At its best it will be unexpected, but overall it will be a disaster if you will need different proxies in one cases and some default in others.

Reply:
I can see this in two ways (and no one being wrong):

1. System administrator (root) should be able to control how anything running on the machine should interact with the rest of the network, therefore, if ENV is set, it takes precedence
1. Developer should have total control on its own code, so if proxy is defined, it takes precedence

If we are inclined to favor the developer, I can suggest we do the following instead:

1. if at least one of the following are filled by the user/developer, ignore ENV: `proxy_address`, `proxy_port`, `proxy_username`, `proxy_password`
1. always merge `proxy_headers` (so we preserve it when using values provided by ENV)

Reply:
@brodock see my comments here: https://github.com/httprb/http/pull/578#issuecomment-566072228

This isn't a dichotomy: many environments require both Internet access via an egress proxy, and S2S communication which doesn't go through the egress proxy.

Changing any kind of defaults is a semver breaking change that could have impacts on real-world environments where this library is currently deployed (I say this rather empirically, as this library is heavily utilized in a large enterprise company where I used to work).

Reply:
I wasn't aware of a use-case where you would have different proxy configurations like that.

So direction is to implement this new behaviour in `via_default_proxy`? 
As it is a chainable method, what should happen if you have `HTTP.via(...).via_default_proxy...` ?

Reply:
I believe the behavior for that sort of thing elsewhere is to overwrite the previous configuration (or rather, split/dup it and then set the new value)

Reply:
So in my example above, assuming:

```ruby
ENV["http_proxy"] = "http://proxy1.example.com:9090"
```

```ruby
HTTP.via("proxy2.example.com", 8080).via_default_proxy()...
# proxy_address: proxy1.example.com
# proxy_port: 9090
```

but keeping credentials from `via`, wouldn't make sense, and we should unset those if default detects anything (as the credentials there are for the first proxy not the second)

```ruby
HTTP.via("proxy2.example.com", 8080, "douglas", "adams").via_default_proxy()...
```

and the third case would be to preserve the headers:

```ruby
HTTP.via("proxy2.example.com", 8080, "douglas", "adams", {"X-Custom-Header" => "some value"}).via_default_proxy()...
```

Reply:
I tried to implement `via_default_proxy` with previously proposed semantics and it's not possible as we need the request_uri in advance which is not garanteed to be available during chainable definition.

I tried an alternative: `with_system_proxy`, this defines an extra proxy_param: `proxy_auto_detection: true`, which gets evaluated when building the request.

It will first try to check if user defined proxy is "available?", if that fails, it wil check whether auto_detection is set and try to use it, falling back to unset proxy when that also fails.

(still need to cover auto detection flag with specs)